src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+      val numMismatches = if (matchReverseComplement) {
+        if (!sampleBarcode.isInstanceOf[CustomSampleBarcode]) unreachable("Cannot match the reverse complement unless the barcode is of type 'CustomSampleBarcode'.")
+        val reverseMismatches = countMismatches(sampleBarcodeReadBases.getBytes, sampleBarcode.asInstanceOf[CustomSampleBarcode].reverseComplementedBarcodeBytes)
+        require(reverseMismatches != 0 || forwardMismatches != 0) // otherwise they are reverse complements of each other
+        if (forwardMismatches < reverseMismatches) forwardMismatches
+        else reverseMismatches
+      }
+      else {
+        forwardMismatches
+      }
+      (sample.sampleOrdinal, numMismatches)
+    }
+
+    /** Searches for a matching sample given the observed barcode.  Returns -1 if no match is found, given the
+      * sample sheet and matching parameters */
+    private def getSampleOrdinalFromSampleBarcode(sampleBarcodeReadBases: String, sampleSheet: SampleSheet): Option[Int] = {
+      val numNoCalls: Int = sampleBarcodeReadBases.getBytes.count(base => SequenceUtil.isNoCall(base))
+      val (bestSampleOrdinal, bestMismatches, secondBestMismatches) = if (numNoCalls <= maxNoCalls) {
+        sampleSheet.map { sample => sampleOrdinalAndMismatches(sample, sampleBarcodeReadBases) }
+          .toList.sortBy(_._2).take(2) match {
+            case Nil                              => (-1,           Integer.MAX_VALUE, Integer.MAX_VALUE)
+            case List(bestTuple)                  => (bestTuple._1, bestTuple._2,      Integer.MAX_VALUE)
+            case List(bestTuple, secondBestTuple) => (bestTuple._1, bestTuple._2,      secondBestTuple._2)
+          }
+      }
+      else {
+        (-1, Integer.MAX_VALUE, Integer.MAX_VALUE)
+      }
+      // Make sure we are within the parameter limits and update barcode metrics if necessary
+      if (maxMismatches < bestMismatches || maxNoCalls < numNoCalls || (secondBestMismatches - bestMismatches) < minMismatchDelta) {
+        noMatchBarcodeMetric.reads +=  1
+        noMatchBarcodeMetric.pf_reads += 1
+        None
+      }
+      else {
+        val sampleBarcode = sampleSheet.get(bestSampleOrdinal - 1).sampleBarcode.getOrElse(unreachable(s"Sample barcode required for sample"))
+        val bestBarcodeMetric: SampleBarcodeMetric = this.barcodeToMetrics.getOrElse(sampleBarcode.concatenatedBarcode, unreachable("No metric for sample"))
+        bestBarcodeMetric.reads += 1
+        bestBarcodeMetric.pf_reads += 1
+        if (bestMismatches == 0) {
+          bestBarcodeMetric.perfect_matches += 1
+          bestBarcodeMetric.pf_perfect_matches += 1
+        }
+        else if (bestMismatches == 1) {
+          bestBarcodeMetric.one_mismatch_matches += 1
+          bestBarcodeMetric.pf_one_mismatch_matches += 1
+        }
+        Some(bestSampleOrdinal)
+      }
+    }
+
+    /** Creates a SAM record (always paired) */
```

Comment:
Can we not handle unpaired reads?  It seems like we really should if this is going to be a generally useful tool.